src/main/java/org/prebid/server/execution/HttpResponseSender.java

```diff
@@ -0,0 +1,150 @@
+        final HttpResponseSender responder = new HttpResponseSender();
+
+        responder.routingContext = routingContext;
+        responder.logger = logger;
+
+        return responder;
+    }
+
+    public HttpResponseSender exceptionHandler(Handler<Throwable> exceptionHandler) {
+        this.exceptionHandler = exceptionHandler;
+        return this;
+    }
+
+    /**
+     * Sets HTTP response status, default is {@link HttpResponseStatus#OK} if not set.
+     */
+    public HttpResponseSender status(HttpResponseStatus status) {
+        this.status = status;
+        return this;
+    }
+
+    public HttpResponseSender headers(Map<CharSequence, CharSequence> headers) {
+        this.headers = headers;
+        return this;
+    }
+
+    public HttpResponseSender addHeader(CharSequence name, CharSequence value) {
+        if (headers == null) {
+            headers = new HashMap<>();
+        }
+        headers.put(name, value);
+        return this;
+    }
+
+    public HttpResponseSender body(String body) {
+        this.body = body;
+        return this;
+    }
+
+    public HttpResponseSender body(Buffer body) {
+        this.bodyAsBuffer = body;
+        return this;
+    }
+
+    /**
+     * Sends HTTP response to the client.
+     * <p>
+     * Returns true if response was sent, otherwise false.
+     */
+    public boolean send() {
+        if (body != null) {
+            return respondWith(body, routingContext.response()::end);
+        }
+
+        if (bodyAsBuffer != null) {
+            return respondWith(bodyAsBuffer, routingContext.response()::end);
+        }
+
+        return respondWith(null, emptyConsumer());
+    }
+
+    /**
+     * Sends file response to the client.
+     */
+    public void sendFile(String filename) {
+        respondWith(filename, routingContext.response()::sendFile);
+    }
+
+    private <T> boolean respondWith(T data, Consumer<T> responseSender) {
+        if (connectionClosed()) {
+            return false;
+        }
+
+        final HttpServerResponse response = routingContext.response();
+
+        if (exceptionHandler != null) {
+            response.exceptionHandler(exceptionHandler);
+        }
+
+        if (status != null) {
+            response.setStatusCode(status.code());
+            response.setStatusMessage(status.reasonPhrase());
+        }
+
+        if (MapUtils.isNotEmpty(headers)) {
+            headers.forEach(response::putHeader);
+        }
+
+        if (data != null) {
+            responseSender.accept(data);
+        } else {
+            response.end();
+        }
+        return true;
+    }
+
+    /**
+     * Return true if client has closed underlying connection, otherwise false.
+     */
+    private boolean connectionClosed() {
+        final boolean connectionClosed = routingContext.response().closed();
+
+        if (connectionClosed) {
+            ConditionalLoggerFactory.getOrCreate(logger)
```

Comment:
I'm not quite sure it's necessary for callers to provide their logger to this class. If the only reason for doing that to distinguish log messages by endpoint, it would be much simpler and clearer to accept some qualifier (for example enum value designating an endpoint) and use it in the log message. In this case there will be no need in ConditionalLoggerFactory and this class will use its own logger for all conditional logging.

Reply:
ok, 
also, let's make HttpResponseSender to be component, so metrics can injected.